### PR TITLE
feat: tests

### PR DESCRIPTION
**Pull request by johnnyonline**
Coverage:
```
// BEFORE
// | File                                           | % Lines          | % Statements     | % Branches    | % Funcs        |
// | src/StakingNode.sol                            | 8.65% (9/104)    | 9.15% (14/153)   | 0.00% (0/26)  | 18.52% (5/27)  |
// | src/StakingNodesManager.sol                    | 8.40% (10/119)   | 11.03% (16/145)  | 2.78% (1/36)  | 15.38% (4/26)  |
// | src/WithdrawalQueueManager.sol                 | 11.76% (8/68)    | 9.68% (9/93)     | 0.00% (0/22)  | 11.11% (2/18)  |
// | src/ynETHRedemptionAssetsVault.sol             | 33.33% (8/24)    | 33.33% (9/27)    | 12.50% (1/8)  | 22.22% (2/9)   |

// AFTER
// | src/StakingNode.sol                            | 94.23% (98/104)    | 94.12% (144/153)   | 65.38% (17/26)   | 100.00% (27/27)  |
// | src/StakingNodesManager.sol                    | 94.12% (112/119)   | 95.17% (138/145)   | 69.44% (25/36)   | 100.00% (26/26)  |
// | src/WithdrawalQueueManager.sol                 | 88.24% (60/68)     | 89.25% (83/93)     | 77.27% (17/22)   | 77.78% (14/18)   |
// | src/ynETHRedemptionAssetsVault.sol             | 83.33% (20/24)     | 85.19% (23/27)     | 75.00% (6/8)     | 100.00% (9/9)    |
```

### CODE REVIEW

**Comment by johnnyonline** [comment]
1. make sure that lists are the same length at `claimWithdrawals(uint256[] calldata tokenIds, address[] calldata receivers)`
2. `creationBlock` from `WithdrawalRequest` struct is not used
3. consider adding a limit to `setSecondsToFinalization()`, and emitting an event
4. not initializing `REDEMPTION_ASSET_WITHDRAWER_ROLE` role admin will render the `withdrawSurplusRedemptionAssets` function useless
5. `queueManager.withdrawSurplusRedemptionAssets()` and `redemptionAssetsVault.withdrawRedemptionAssets()` will always fail because of the `ynETH.processWithdrawnETH()` call - which is gated to `onlyStakingNodesManager`
6. `ReentrancyGuardUpgradeable` not used in `ynETHRedemptionAssetsVault.sol`
7. make `_paused` public in `ynETHRedemptionAssetsVault.sol`

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+    // ------------------------------------------
+    // Withdrawal flow tests
+    // ------------------------------------------
+
+    function testVerifyWithdrawalCredentials() public {
+        if (!isHolesky) return;
+
+        ValidatorProofs memory _validatorProofs = getWithdrawalCredentialParams();
+
+        vm.expectRevert(bytes4(keccak256("NotStakingNodesOperator()")));
```
Comment (danoctavian) [inline test/scenarios/fork/ynETHWithdrawals.t.sol:73]: do not mix error cases with success cases in the same test.

They are separate tests

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+        );
+
+        assertEq(stakingNode.unverifiedStakedETH(), _unverifiedStakedETH, "testVerifyWithdrawalCredentials: E0"); // validator already verified withdrawal credentials
+    }
+
+    function testQueueWithdrawal() public {
+        if (!isHolesky) return;
+
+        // uint256 _amount = withdrawalAmount;
+        // fuzzing here will fail on too many inputs rejected
```
Comment (danoctavian) [inline test/scenarios/fork/ynETHWithdrawals.t.sol:100]: making fuzzing work with these tests that verifyWithdrawalCredentials may not be feasible.

The amount of logic executed on forked state is very high.

Fuzz testing works well on freshly deployed code that doesn't interact with a ton of pre-existing forked state.

So not surprised by this result.

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+        // vm.assume(_amount > 1e9); // _amount > 1 gwei
+        // vm.assume(_amount % 1e9 == 0); // _amount must be a whole Gwei amount
+        // vm.assume(_amount <= uint256(IEigenPodManager(stakingNodesManager.eigenPodManager()).podOwnerShares(address(stakingNode))));
+
+        testVerifyWithdrawalCredentials();
+
+        vm.expectRevert(bytes4(keccak256("NotStakingNodesOperator()")));
```
Comment (danoctavian) [inline test/scenarios/fork/ynETHWithdrawals.t.sol:107]: separate test case

**Comment by danoctavian** [comment]
To fix the conflicts with base branch, please use *rebase* @johnnyonline 

if you need help, with rebasing certain parts, let us know

File: test/utils/StakingNodeTestBase.sol [hunk 4]
```diff
@@ -197,6 +198,9 @@ contract StakingNodeTestBase is ScenarioBaseTest, ProofParsingV1 {
 
         vm.roll(block.number + delegationManager.minWithdrawalDelayBlocks() + 1);
 
+        vm.expectRevert(bytes4(keccak256("NotStakingNodesOperator()")));
```
Comment (danoctavian) [inline test/utils/StakingNodeTestBase.sol:201]: don't add specific testing of error cases within other tests.

Tests these in separate tests whose name describes what's being tested

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+        _queuedSharesAmountBefore = stakingNode.queuedSharesAmount();
+        uint256 _balanceBefore = address(stakingNode).balance;
+        uint256 _withdrawnValidatorPrincipalBefore = stakingNode.withdrawnValidatorPrincipal();
+        completeQueuedWithdrawals(stakingNode, withdrawalAmount);
+        assertEq(address(stakingNode).balance - _balanceBefore, withdrawalAmount, "testQueueWithdrawal: E1");
+        assertEq(stakingNode.queuedSharesAmount(), _queuedSharesAmountBefore - withdrawalAmount, "testQueueWithdrawal: E2");
+        assertEq(stakingNode.withdrawnValidatorPrincipal(), _withdrawnValidatorPrincipalBefore + withdrawalAmount, "testQueueWithdrawal: E3");
+    }
+
+    function testRequestWithdrawal(uint256 _amount) public {
```
Comment (danoctavian) [inline test/scenarios/fork/ynETHWithdrawals.t.sol:126]: this test says requestWithdrawal but tests all the way to claim.

We need to make tests that test a specific thing, and that thing is roughly described in the name.

There can be a test here that simply does requestWithdrawal and does all the necessary assertions.

The following one does the claimWithdrawal and focuses on asserting things for that.

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+        {
+            uint256 _timestamp = block.timestamp;
+            vm.warp(block.timestamp + ynETHWithdrawalQueueManager.secondsToFinalization());
+            vm.prank(user);
+            // vm.expectRevert(abi.encodeWithSelector(InsufficientBalance.selector, address(ynETHWithdrawalQueueManager).balance, _amount));
+            // fails with above revert reason, but _amount is a little less because of precision loss
+            // ynETHRedemptionAssetsVaultInstance.availableRedemptionAssets(address(0));
+            if (ynETHRedemptionAssetsVaultInstance.availableRedemptionAssets(address(0)) < _amount) {
+                vm.expectRevert();
+                ynETHWithdrawalQueueManager.claimWithdrawal(tokenId, receiver);
```
Comment (danoctavian) [inline test/scenarios/fork/ynETHWithdrawals.t.sol:181]: this verification ere related to claimWithdrawal does not belong in this test, and also the revert should be written explicitly 

with selector and params. 

We don't want to use empty vm.expectRevert()

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+            ynETHRedemptionAssetsVaultInstance.transferRedemptionAssets(_to, _amount);
+
+            vm.prank(address(ynETHWithdrawalQueueManager));
+            ynETHRedemptionAssetsVaultInstance.transferRedemptionAssets(_to, _amount);
+            assertEq(address(ynETHRedemptionAssetsVaultInstance).balance, _balanceBefore - _amount, "testTransferRedemptionAssets: E0");
+            assertEq(address(_to).balance - _toBalanceBefore, _amount, "testTransferRedemptionAssets: E1");
+        }
+    }
+
+    // function testWithdrawRedemptionAssets(uint256 _amount) public {
```
Comment (danoctavian) [inline test/scenarios/fork/ynETHWithdrawals.t.sol:485]: end state of the PR cannot contain commented out code, unless it's a draft.

We won't be merging commented out code

We need to clarify if this is a valid test and it needs to be fixed up, or delete it

File: test/scenarios/fork/ynETHWithdrawals.t.sol [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+        completeQueuedWithdrawals(stakingNode, withdrawalAmount);
+        assertEq(address(stakingNode).balance - _balanceBefore, withdrawalAmount, "testQueueWithdrawal: E1");
+        assertEq(stakingNode.queuedSharesAmount(), _queuedSharesAmountBefore - withdrawalAmount, "testQueueWithdrawal: E2");
+        assertEq(stakingNode.withdrawnValidatorPrincipal(), _withdrawnValidatorPrincipalBefore + withdrawalAmount, "testQueueWithdrawal: E3");
+    }
+
+    function testRequestWithdrawal(uint256 _amount) public {
+        if (!isHolesky) return;
+
+        vm.assume(_amount > 0 && _amount < 10_000 ether);
```
[the remark this answers is not in the document]
Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol:129]: which one is the other one?

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 5 above)
Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol:126]: hmm it tests only the queueing and request part

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 7 above)
Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol:485]: code needs to be fixed in order for those to pass...

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (danoctavian) [reply test/scenarios/fork/ynETHWithdrawals.t.sol:129]: https://github.com/yieldnest/yieldnest-protocol/pull/124/files#diff-c0ae200d624fb99b79d9b6f857fce28bb69f2dd1ae816ff43c52bf9be9b7c0c2R100

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 5 above)
Comment (danoctavian) [reply test/scenarios/fork/ynETHWithdrawals.t.sol:126]: it has this call to claim at the end:
https://github.com/yieldnest/yieldnest-protocol/pull/124/files#r1671790819

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 7 above)
Comment (danoctavian) [reply test/scenarios/fork/ynETHWithdrawals.t.sol:485]: Leave it as a failing test then to show case the test failure

Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol]: fixed in https://github.com/yieldnest/yieldnest-protocol/pull/124#discussion_r1671918664.

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 6 above)
Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol]: fixed in https://github.com/yieldnest/yieldnest-protocol/pull/124/commits/6bf949f723b721a6eac02e32dbbf63fae562bbf5.

File: test/utils/StakingNodeTestBase.sol (hunk 4 above)
Comment (johnnyonline) [reply test/utils/StakingNodeTestBase.sol:201]: fixed in https://github.com/yieldnest/yieldnest-protocol/pull/124/commits/e40cb6f14cb1feb306eafa7a9ba859b03944625e.

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 3 above)
Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol]: fixed in https://github.com/yieldnest/yieldnest-protocol/pull/124/commits/e40cb6f14cb1feb306eafa7a9ba859b03944625e.

File: test/scenarios/fork/ynETHWithdrawals.t.sol (hunk 1 above)
Comment (johnnyonline) [reply test/scenarios/fork/ynETHWithdrawals.t.sol]: fixed in https://github.com/yieldnest/yieldnest-protocol/pull/124/commits/e40cb6f14cb1feb306eafa7a9ba859b03944625e.

**Pull request merged**